migration/migrator/migrations/course/20220612122726_gradeable_specific_anon_ids.py

```diff
@@ -0,0 +1,62 @@
+"""Migration for a given Submitty course database."""
+import json
+from collections import OrderedDict
+
+def up(config, database, semester, course):
+    """
+    Run up migration.
+
+    :param config: Object holding configuration details about Submitty
+    :type config: migrator.config.Config
+    :param database: Object for interacting with given database for environment
+    :type database: migrator.db.Database
+    :param semester: Semester of the course being migrated
+    :type semester: str
+    :param course: Code of course being migrated
+    :type course: str
+    """
+    database.execute(
+        """
+        CREATE TABLE IF NOT EXISTS gradeable_anon (
+            user_id character varying NOT NULL,
+            g_id character varying(255) NOT NULL,
+            anon_id character varying(255),
+            FOREIGN KEY(user_id)
```

Comment:
Should we also add a UNIQUE key here of g_id and anon_d? I know one of the scripts makes sure to not repeat an anon_id but I don't see how it could be harmful to make our database also guarantee it.

Reply:
I don't think `g_id` is unique, all the users in `gradeable_anon` have same `g_id`. Also, after modifying User.php to send `$g_id` i.e. `$anon_ids = $this->core->getQueries()->getAllAnonIdsByGradeable($g_id)`. `anon_id` is not guaranteed to be unique when whole table is considered, it is only unique in the context of a particular gradeable, practically it can be expected to be unique across the table but not theoretically.

Reply:
I was referring that there can be a unique key/constraint consisting of BOTH g_id and anon_id. I know there can definitely be multiple g_id's and possibly multiple anon_id's but there should never be a multiple rows where the g_id and anon_id are both equal.

Reply:
Here is an example of this https://github.com/Submitty/Submitty/blob/2590ba5ea747dda50b5e1f7e862de9f582292e72/migration/migrator/data/course_tables.sql#L1853

Reply:
Oh, okay, my bad, will do it.